docs/wallet/chain-query-api.md

```diff
@@ -0,0 +1,201 @@
+// or writing it to disk
+  val exampleProcessTx = (tx: Transaction) =>
+    println(s"Received tx: ${tx.txIdBE}")
+
+val exampleProcessBlock = (block: Block) =>
+    println(s"Received block: ${block.blockHeader.hashBE}")
+
+val exampleProcessFilter =
+    (blockHash: DoubleSha256Digest, filter: GolombFilter) =>
+      println(s"Received filter: ${blockHash.flip.hex} ${filter.hash.flip.hex}")
+
+val exampleCallbacks =
+    createCallbacks(exampleProcessTx, exampleProcessFilter, exampleProcessBlock)
+
+// Here is where we are defining our actual chain api, Ideally this could be it's own class
+// but for the examples sake we will keep it small.
+  val chainApi = new ChainQueryApi {
+
+    /** Gets the height of the given block */
+    override def getBlockHeight(
+        blockHash: DoubleSha256DigestBE): Future[Option[Int]] = {
+      bitcoind.getBlock(blockHash).map(block => Some(block.height))
+    }
+
+    /** Gets the hash of the block that is what we consider "best" */
+    override def getBestBlockHash(): Future[DoubleSha256DigestBE] = {
+      bitcoind.getBestBlockHash
+    }
+
+    /** Gets number of confirmations for the given block hash */
+    override def getNumberOfConfirmations(
+        blockHash: DoubleSha256DigestBE): Future[Option[Int]] = {
+      for {
+        tip <- bitcoind.getBlockCount
+        block <- bitcoind.getBlock(blockHash)
+      } yield {
+        Some(tip - block.height + 1)
+      }
+    }
+
+    /** Gets the number of compact filters in the database */
+    override def getFilterCount: Future[Int] = {
+      // since bitcoind should have the filter for
+      // every block we can just return the block height
+      bitcoind.getBlockCount
+    }
+
+    /** Returns the block height of the given block stamp */
+    override def getHeightByBlockStamp(blockStamp: BlockStamp): Future[Int] = {
+      blockStamp match {
+        case blockHeight: BlockStamp.BlockHeight =>
+          Future.successful(blockHeight.height)
+        case blockHash: BlockStamp.BlockHash =>
+          getBlockHeight(blockHash.hash).map(_.get)
+        case blockTime: BlockStamp.BlockTime =>
+          Future.failed(new RuntimeException(s"Not implemented: $blockTime"))
+      }
+    }
+
+    override def getFiltersBetweenHeights(
+        startHeight: Int,
+        endHeight: Int): Future[Vector[FilterResponse]] = {
+      val filterFs = startHeight
+        .until(endHeight)
+        .map { height =>
+          for {
+            hash <- bitcoind.getBlockHash(height)
+            filter <- bitcoind.getBlockFilter(hash, FilterType.Basic)
+          } yield {
+            FilterResponse(filter.filter, hash, height)
+          }
+        }
+        .toVector
+
+      Future.sequence(filterFs)
+    }
+  }
+
+// Finally, we can initialize our wallet with our own node api
+  val wallet =
+    Wallet(keyManager = keyManager, nodeApi = nodeApi, chainQueryApi = chainApi)
```

Comment:
same thing as I said in https://github.com/bitcoin-s/bitcoin-s/pull/1205#discussion_r389860261 , just trigger one of the callbacks so someone can see how this is executed. Eventually I would like to move these to into scripts that can be run on their own. See #864 for what that would look like.